core/unit_test/TestRangePolicyConstructors.hpp

```diff
@@ -0,0 +1,391 @@
+      using schedule_type   = typename policy_t::schedule_type;
+      using work_tag        = typename policy_t::work_tag;
+
+      static_assert(std::is_same<execution_space, Kokkos::DefaultExecutionSpace>::value, "");
+      static_assert(std::is_same<index_type, long>::value, "");
+      static_assert(std::is_same<schedule_type, Kokkos::Schedule<Kokkos::Dynamic>>::value, "");
+      static_assert(std::is_same<work_tag, SomeTag>::value, "");
+    }
+  }
+
+  void test_runtime_parameters() {
+    using policy_t     = Kokkos::RangePolicy<>;
+    using index_t      = policy_t::index_type;
+    index_t work_begin = 5;
+    index_t work_end   = 15;
+    index_t chunk_size = 10;
+    {
+      policy_t p(work_begin, work_end);
+      ASSERT_EQ(p.begin(), work_begin);
+      ASSERT_EQ(p.end(),   work_end);
+    }
+    {
+      policy_t p(ExecutionSpace(), work_begin, work_end);
+      ASSERT_EQ(p.begin(), work_begin);
+      ASSERT_EQ(p.end(),   work_end);
+    }
+    {
+      policy_t p(work_begin, work_end, Kokkos::ChunkSize(chunk_size));
+      ASSERT_EQ(p.begin(),      work_begin);
+      ASSERT_EQ(p.end(),        work_end);
+      ASSERT_EQ(p.chunk_size(), chunk_size);
+    }
+    {
+      policy_t p(ExecutionSpace(), work_begin, work_end,
+                 Kokkos::ChunkSize(chunk_size));
+      ASSERT_EQ(p.begin(),      work_begin);
+      ASSERT_EQ(p.end(),        work_end);
+      ASSERT_EQ(p.chunk_size(), chunk_size);
+    }
+    {
+      policy_t p;
+      ASSERT_EQ(p.begin(),      index_t(0));
+      ASSERT_EQ(p.end(),        index_t(0));
+      ASSERT_EQ(p.chunk_size(), index_t(0));
+
+      p = policy_t(work_begin, work_end, Kokkos::ChunkSize(chunk_size));
+      ASSERT_EQ(p.begin(),      work_begin);
+      ASSERT_EQ(p.end(),        work_end);
+      ASSERT_EQ(p.chunk_size(), chunk_size);
+    }
+    {
+      policy_t p1(work_begin, work_end, Kokkos::ChunkSize(chunk_size));
+      policy_t p2(p1);
+      ASSERT_EQ(p1.begin(),      p2.begin());
+      ASSERT_EQ(p1.end(),        p2.end());
+      ASSERT_EQ(p1.chunk_size(), p2.chunk_size());
+    }
+  }
+
+  void test_worktag() {
+    struct WorkTag1 {};
+    struct WorkTag2 {};
+
+    using policy_t_no_worktag = Kokkos::RangePolicy<>;
+    using policy_t_worktag1 = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_no_worktag, WorkTag1>;
+    using policy_t_worktag2 = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_worktag1, WorkTag2>;
+
+    policy_t_no_worktag p0;
+    static_assert(std::is_void<typename decltype(p0)::work_tag>::value, "");
+
+    auto p1 = policy_t_worktag1(p0);
+    static_assert(std::is_same<typename decltype(p1)::work_tag, WorkTag1>::value, "");
+
+    auto p2 = policy_t_worktag2(p1);
+    static_assert(std::is_same<typename decltype(p2)::work_tag, WorkTag2>::value, "");
+
+    // NOTE line directly below does not currently compile
+    //using policy_t_void_tag = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_no_worktag, void>;
```

Comment:
Is this something we eventually want to work, or want to have documented that it doesn't work? I just copied this over from the other tests file, but can remove if it is not needed?

Reply:
Why doesn't it work? What's the error message?

Reply:
We get a static_assert fail in `core/src/traits/Kokkos_PolicyTraitAdaptor.hpp:123`
```
  static_assert(PolicyTraitMatcher<TraitSpec, NewTrait>::value, "");
```
The `PolicyTraitMatcher` struct looks like 
```
template <class TraitSpec, class Trait, class Enable = void>
struct PolicyTraitMatcher : std::false_type {};

template <class TraitSpec, class Trait>
struct PolicyTraitMatcher<
    TraitSpec, Trait,
    std::enable_if_t<
        TraitSpec::template trait_matches_specification<Trait>::value>>
    : std::true_type {};
```
Is it complaining that it is expecting a `struct` but is given `void`?